include/sparrow/buffer.hpp

```diff
@@ -18,454 +18,870 @@
+        : base_type(check_init_length(n, a), a)
+    {
+        get_data().p_end = fill_initialize(get_data().p_begin, n, v, get_allocator());
+    }
+
+    template <class T>
+    template <allocator A>
+    buffer<T>::buffer(pointer p, size_type n, const A& a)
+        : base_type(p, check_init_length(n, a), a)
+    {
+    }
+
+    template <class T>
+    template <allocator A>
+    buffer<T>::buffer(std::initializer_list<value_type> init, const A& a)
+        : base_type(check_init_length(init.size(), a), a)
+    {
+        get_data().p_end = copy_initialize
+        (
+            init.begin(),
+            init.end(),
+            get_data().p_begin,
+            get_allocator()
+        );
+    }
+
+    template <class T>
+    template <class It, allocator A>
+    buffer<T>::buffer(It first, It last, const A& a)
+        : base_type(check_init_length(std::distance(first, last), a), a)
+    {
+        get_data().p_end = copy_initialize
+        (
+            first,
+            last,
+            get_data().p_begin,
+            get_allocator()
+        );
+    }
+
+    template <class T>
+    buffer<T>::~buffer()
+    {
+        destroy(get_data().p_begin, get_data().p_end, get_allocator());
+    }
+
+    template <class T>
+    buffer<T>::buffer(const buffer& rhs)
+        : base_type(rhs.size(), rhs.get_allocator())
+    {
+        get_data().p_end = copy_initialize
+        (
+            rhs.begin(),
+            rhs.end(),
+            get_data().p_begin,
+            get_allocator()
+        );
+    }
 
     template <class T>
-    buffer_base<T>::buffer_base(pointer p, size_type n)
-        : p_data(p)
-        , m_size(n)
+    template <allocator A>
+    buffer<T>::buffer(const buffer& rhs, const A& a)
+        : base_type(rhs.size(), a)
     {
+        get_data().m_end = copy_initialize
+        (
+            rhs.begin(),
+            rhs.end(),
+            get_data().p_begin,
+            get_allocator()
+        );
     }
 
     template <class T>
-    void buffer_base<T>::reset(pointer p, size_type n)
+    template <allocator A>
+    buffer<T>::buffer(buffer&& rhs, const A& a)
+        : base_type(a)
     {
-        p_data = p;
-        m_size = n;
+        if (rhs.get_allocator() == get_allocator())
+        {
+            get_data() = std::move(rhs.m_data);
+        }
+        else if (!rhs.empty())
+        {
+            this->create_storage(rhs.size());
+            get_data().p_end = copy_initialize
+            (
+                rhs.begin(),
+                rhs.end(),
+                get_data().p_begin,
+                get_allocator()
+            );
+            rhs.clear();
+        }
     }
 
     template <class T>
-    bool buffer_base<T>::empty() const noexcept
+    buffer<T>& buffer<T>::operator=(const buffer& rhs)
     {
-        return size() == size_type(0);
+        if (std::addressof(rhs) != this)
+        {
+            // We assume that any_allocator nevers propagates on assign
```

Comment:
What do you mean by this comment exactly?
```suggestion
            // We assume that any_allocator never propagates on assign.
```

Reply:
We do not assign allocators from a buffer instance to another one.